python/lib/modeldata/dmod/modeldata/data/object_store_dataset.py

```diff
@@ -0,0 +1,369 @@
+
+        Returns
+        -------
+        ObjectWriteResult
+        """
+        if do_checks:
+            if not file.exists():
+                raise RuntimeError("Cannot push non-existing file {} to bucket {}".format(str(file), bucket_name))
+            elif not file.is_file():
+                raise RuntimeError("Cannot push non-regular file {} to bucket {}".format(str(file), bucket_name))
+            elif not file.is_relative_to(bucket_root):
+                msg = "Cannot push {} to bucket {} when provided bad or non-relative bucket root {}"
+                raise RuntimeError(msg.format(str(file), bucket_name, str(bucket_root)))
+        file_rel_root = file.relative_to(bucket_root)
+        object_name = self._OBJECT_NAME_SEPARATOR.join(file_rel_root.parts)
+        return self._client.fput_object(bucket_name=bucket_name, object_name=object_name, file_path=str(file))
+
+    # TODO: might need to add the threading stuff in this function when ready to add it
+    def _push_files(self, bucket_name: str, dir_path: Path, recursive: bool = True, bucket_root: Optional[Path] = None,
+                    do_checks: bool = True):
+        """
+        Push the file contents of the given directory to the provided bucket.
+
+        Parameters
+        ----------
+        bucket_name : str
+            The name of the existing bucket to push to.
+        dir_path : Path
+            The name of the existing directory containing files to push to the bucket.
+        recursive : bool
+            Whether contents of nested directories, their inner directory contents, etc., should all be pushed, as
+            opposed to only regular files that are immediate children of ``dir_path`` (``True`` by default)
+        bucket_root : Optional[Path]
+            The directory level that corresponds to the bucket's root level, for object naming purposes.
+        do_checks : bool
+            Whether to do sanity checks on the local file and bucket root (``True`` by default).
+        """
+        if do_checks:
+            if not dir_path.exists():
+                msg = "Cannot push files from non-existing directory {} to bucket {}"
+                raise RuntimeError(msg.format(str(dir_path), bucket_name))
+            elif not dir_path.is_dir():
+                msg = "Cannot push directory contents from non-directory {} to bucket {}"
+                raise RuntimeError(msg.format(str(dir_path), bucket_name))
+        if bucket_root is None:
+            bucket_root = dir_path
+        # First take care of immediate
+        for file in [f for f in dir_path.iterdir() if f.is_file()]:
+            self._push_file(bucket_name, file, bucket_root, do_checks=False)
+        if recursive:
+            for directory in [d for d in dir_path.iterdir() if d.is_dir()]:
+                self._push_files(bucket_name, directory, recursive, bucket_root, do_checks=False)
+
+    def add_data(self, dataset_name: str, **kwargs) -> bool:
```

Comment:
I considered that, but I decided against it.  

The object store dataset and its manager support a simulated directory structure by encoding the structure into the object names.  To make that work, there has to be a defined "bucket root" directory when adding files, with only a files ancestor directory structure below this bucket root level being encoded.  

If we add all the files in a directory, we can safely assume they all should be added relative to the same bucket root.  But we couldn't make this kind of assumption for an arbitrary list of files, so we'd also need a list of bucket root values (unless we don't care about losing the original directory structure for each file, but I think we do).  We could require a list of files and bucket roots, but I thought that would make the interface too messy, and that it would be better to leave it to something else to provide that convenience if it was wanted.

I'm be open to talking through this more though if you want (either as part of the PR or for later ones).

Reply:
I was assuming the client didn't need the directory structure if they were adding a list of arbitrary files...if they did, they should pass the `Path` to those files, which would get encoded in the object names and put in the same `root` bucket....